### PR TITLE
Change instance profile if the instance is stopped

### LINKED ISSUE #8262: Error on EC2 instance role change
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v0.11.13
+ provider.aws v2.5.0
```

### Affected Resource(s)

* aws_instance

### Terraform Configuration Files

```hcl
resource "aws_instance" "host0001" {
  ami           = "ami-086063d8c95bfa211"
  instance_type = "t3.medium"
  ebs_optimized = "true"
  subnet_id     = "sg-12345678"

  vpc_security_group_ids = [
    "sg-xxxxxxxx",
  ]

  iam_instance_profile = "my-profile"

  tags {
    Name = "host0001"
  }
}
```

### Debug Output

https://gist.github.com/limitusus/429bef80ba7ffc7af90083c886e048f6

### Panic Output

No Panic

### Expected Behavior

IAM profile change is only supported for "running" instances, so for "stopped" instances, remove the attached profile and then attach a new profile.

### Actual Behavior

terraform tries to switch profile directly regardless of the instance's state and causes error.

### Steps to Reproduce

1. Prepare .tf file for an EC2 instance with profile A
1. `terraform apply` and the instance is created
1. stops the instance
1. Edit .tf file for the EC2 instance with different profile B
1. `terraform apply` again and error reproduces

### References

[AWS document](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#replace-iam-role) says
> To replace the IAM role on an instance that already has an attached IAM role, the instance must be in the running state.

### PR DESCRIPTION
**Pull request by ryndaniels**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Closes #8262

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
* resource/aws_iam_instance: Allows for instance profiles to be changed when instances are in stopped state
```

Output from acceptance testing:

* Updated test, before the fix:
```
$ make testacc TESTARGS="-run=TestAccAWSInstance_instanceProfileChange"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -count 1 -parallel 20 -run=TestAccAWSInstance_instanceProfileChange -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAWSInstance_instanceProfileChange
=== PAUSE TestAccAWSInstance_instanceProfileChange
=== CONT  TestAccAWSInstance_instanceProfileChange
--- FAIL: TestAccAWSInstance_instanceProfileChange (288.59s)
    testing.go:635: Step 4 error: After applying this step, the plan was not empty:
        
        DIFF:
        
        UPDATE: aws_instance.test
          ami:                                       "ami-054a6c7ab667eefef" => "ami-054a6c7ab667eefef"
          arn:                                       "arn:aws:ec2:us-west-2:187416307283:instance/i-01556051a498e2f46" => "arn:aws:ec2:us-west-2:187416307283:instance/i-01556051a498e2f46"
          associate_public_ip_address:               "false" => "false"
          availability_zone:                         "us-west-2c" => "us-west-2c"
          cpu_core_count:                            "1" => "1"
          cpu_threads_per_core:                      "1" => "1"
          credit_specification.#:                    "0" => "0"
          disable_api_termination:                   "false" => "false"
          ebs_block_device.#:                        "0" => "0"
          ebs_optimized:                             "false" => "false"
          ephemeral_block_device.#:                  "0" => "0"
          get_password_data:                         "false" => "false"
          iam_instance_profile:                      "tf-testacc-instance-okvewpoey3sq" => "tf-testacc-instance-btxxqxelrysv"
          id:                                        "i-01556051a498e2f46" => "i-01556051a498e2f46"
          instance_state:                            "stopped" => "stopped"
          instance_type:                             "m1.small" => "m1.small"
          ipv6_address_count:                        "0" => "0"
          ipv6_addresses.#:                          "0" => "0"
          key_name:                                  "" => ""
          monitoring:                                "false" => "false"
          network_interface.#:                       "0" => "0"
          password_data:                             "" => ""
          placement_group:                           "" => ""
          primary_network_interface_id:              "eni-0c28c1cd080d3da24" => "eni-0c28c1cd080d3da24"
          private_dns:                               "ip-172-31-39-108.us-west-2.compute.internal" => "ip-172-31-39-108.us-west-2.compute.internal"
          private_ip:                                "172.31.39.108" => "172.31.39.108"
          public_dns:                                "" => ""
          public_ip:                                 "" => ""
          root_block_device.#:                       "1" => "1"
          root_block_device.0.delete_on_termination: "true" => "true"
          root_block_device.0.encrypted:             "false" => "false"
          root_block_device.0.iops:                  "0" => "0"
          root_block_device.0.kms_key_id:            "" => ""
          root_block_device.0.volume_id:             "vol-070f787eef52c226b" => "vol-070f787eef52c226b"
          root_block_device.0.volume_size:           "2" => "2"
          root_block_device.0.volume_type:           "standard" => "standard"
          security_groups.#:                         "1" => "1"
          security_groups.0:                         "default" => "default"
          source_dest_check:                         "true" => "true"
          subnet_id:                                 "subnet-0ba7e7e3b57ece13c" => "subnet-0ba7e7e3b57ece13c"
          tags.Name:                                 "tf-testacc-instance-btxxqxelrysv" => "tf-testacc-instance-btxxqxelrysv"
          tenancy:                                   "default" => "default"
          vpc_security_group_ids.#:                  "1" => "1"
          vpc_security_group_ids.0:                  "sg-1ca59078" => "sg-1ca59078"
        
        
        
        STATE:
...
FAIL
FAIL    github.com/terraform-providers/terraform-provider-aws/aws       290.088s
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws/internal/flatmap      0.031s [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags 0.579s [no tests to run]
FAIL
make: *** [testacc] Error 1
```

* Updated test, after the fix:
```
make testacc TESTARGS="-run=TestAccAWSInstance_instanceProfileChange"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -count 1 -parallel 20 -run=TestAccAWSInstance_instanceProfileChange -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAWSInstance_instanceProfileChange
=== PAUSE TestAccAWSInstance_instanceProfileChange
=== CONT  TestAccAWSInstance_instanceProfileChange
--- PASS: TestAccAWSInstance_instanceProfileChange (318.98s)
PASS
```

* The other tests, after the fix:
```
--- SKIP: TestAccAWSInstance_inEc2Classic (2.07s)
--- PASS: TestAccAWSInstance_ipv6AddressCountAndSingleAddressCausesError (16.07s)
--- PASS: TestAccAWSInstance_GP2IopsDevice (65.05s)
--- PASS: TestAccAWSInstance_placementGroup (67.97s)
--- PASS: TestAccAWSInstance_GP2WithIopsValue (70.74s)
--- PASS: TestAccAWSInstance_vpc (80.24s)
--- PASS: TestAccAWSInstance_rootInstanceStore (81.31s)
--- PASS: TestAccAWSInstance_NetworkInstanceSecurityGroups (90.12s)
--- PASS: TestAccAWSInstance_inDefaultVpcBySgId (93.18s)
--- PASS: TestAccAWSInstance_blockDevices (95.52s)
--- PASS: TestAccAWSInstance_inDefaultVpcBySgName (98.72s)
--- PASS: TestAccAWSInstance_ipv6_supportAddressCountWithIpv4 (99.08s)
--- PASS: TestAccAWSInstance_ipv6_supportAddressCount (107.67s)
--- PASS: TestAccAWSInstance_NetworkInstanceRemovingAllSecurityGroups (92.95s)
--- PASS: TestAccAWSInstance_EbsBlockDevice_KmsKeyArn (109.97s)
--- PASS: TestAccAWSInstance_userDataBase64 (118.29s)
--- PASS: TestAccAWSInstance_basic (129.89s)
--- PASS: TestAccAWSInstance_disableApiTermination (130.68s)
--- PASS: TestAccAWSInstance_sourceDestCheck (137.17s)
--- PASS: TestAccAWSInstance_NetworkInstanceVPCSecurityGroupIDs (76.32s)
--- PASS: TestAccAWSInstance_volumeTags (90.72s)
--- PASS: TestAccAWSInstance_noAMIEphemeralDevices (179.26s)
--- PASS: TestAccAWSInstance_associatePublicIPAndPrivateIP (88.74s)
--- PASS: TestAccAWSInstance_volumeTagsComputed (108.70s)
--- PASS: TestAccAWSInstance_keyPairCheck (95.22s)
--- PASS: TestAccAWSInstance_primaryNetworkInterfaceSourceDestCheck (76.21s)
--- PASS: TestAccAWSInstance_withIamInstanceProfile (107.46s)
--- PASS: TestAccAWSInstance_primaryNetworkInterface (96.54s)
--- PASS: TestAccAWSInstance_associatePublic_defaultPrivate (77.07s)
--- PASS: TestAccAWSInstance_associatePublic_defaultPublic (75.99s)
--- PASS: TestAccAWSInstance_forceNewAndTagsDrift (115.70s)
--- PASS: TestAccAWSInstance_associatePublic_explicitPublic (74.58s)
--- PASS: TestAccAWSInstance_multipleRegions (238.53s)
--- PASS: TestAccAWSInstance_addSecurityGroupNetworkInterface (113.65s)
--- PASS: TestAccAWSInstance_addSecondaryInterface (120.36s)
--- PASS: TestAccAWSInstance_associatePublic_overridePublic (76.59s)
--- PASS: TestAccAWSInstance_associatePublic_explicitPrivate (86.12s)
--- PASS: TestAccAWSInstance_associatePublic_overridePrivate (77.82s)
--- PASS: TestAccAWSInstance_CreditSpecification_UnspecifiedToEmpty_NonBurstable (68.90s)
--- PASS: TestAccAWSInstance_rootBlockDeviceMismatch (176.78s)
--- PASS: TestAccAWSInstance_creditSpecification_unspecifiedDefaultsToStandard (74.60s)
--- PASS: TestAccAWSInstance_privateIP (198.91s)
--- PASS: TestAccAWSInstance_creditSpecification_standardCpuCredits (86.54s)
--- PASS: TestAccAWSInstance_creditSpecification_unlimitedCpuCredits (84.30s)
--- PASS: TestAccAWSInstance_creditSpecificationT3_standardCpuCredits (68.74s)
--- PASS: TestAccAWSInstance_RootBlockDevice_KmsKeyArn (335.06s)
--- PASS: TestAccAWSInstance_creditSpecification_updateCpuCredits (102.64s)
--- PASS: TestAccAWSInstance_getPasswordData_falseToTrue (148.22s)
--- PASS: TestAccAWSInstance_getPasswordData_trueToFalse (158.78s)
--- PASS: TestAccAWSInstance_UserData_EmptyStringToUnspecified (81.09s)
--- PASS: TestAccAWSInstance_tags (308.18s)
--- PASS: TestAccAWSInstance_creditSpecificationT3_unlimitedCpuCredits (115.30s)
--- PASS: TestAccAWSInstance_creditSpecificationT3_updateCpuCredits (120.05s)
--- PASS: TestAccAWSInstance_UserData_UnspecifiedToEmptyString (91.32s)
--- PASS: TestAccAWSInstance_creditSpecification_isNotAppliedToNonBurstable (209.05s)
--- PASS: TestAccAWSInstance_changeInstanceType (349.25s)
--- PASS: TestAccAWSInstance_disappears (211.16s)
--- PASS: TestAccAWSInstance_CreditSpecification_Empty_NonBurstable (309.21s)
--- PASS: TestAccAWSInstance_creditSpecification_unknownCpuCredits_t3 (299.62s)
--- PASS: TestAccAWSInstance_creditSpecificationT3_unspecifiedDefaultsToUnlimited (299.65s)
--- PASS: TestAccAWSInstance_creditSpecification_unknownCpuCredits_t2 (357.64s)
--- PASS: TestAccAWSInstance_creditSpecification_standardCpuCredits_t2Tot3Taint (367.62s)
--- PASS: TestAccAWSInstance_creditSpecification_unlimitedCpuCredits_t2Tot3Taint (391.93s)
```

### CODE REVIEW

File: aws/resource_aws_instance.go [hunk 1]
```diff
@@ -985,33 +965,62 @@ func resourceAwsInstanceUpdate(d *schema.ResourceData, meta interface{}) error {
...
+					if ec2err, ok := err.(awserr.Error); ok && ec2err.Code() == "InvalidInstanceID.NotFound" {
+						d.SetId("")
+						return nil
+					}
+					return err
+				}
+				instance := resp.Reservations[0].Instances[0]
+
+				if instance.State != nil {
+					if *instance.State.Name == "running" {
```
[the remark this answers is not in the document]
Comment (ryndaniels) [reply aws/resource_aws_instance.go]: Good point - it looks like before, those cases would have fallen into the "running" behavior in the code. I *suspect* that `shutting-down` and `stopping` probably ought to fall into the `stopped` state, so I think I'll group those together and have every other state fall through to the old behavior. That way we'll keep things as consistent as possible, and if people run into issues with other state edge cases, we can address those as they come up. 👍

**Comment by ryndaniels** [comment]
Updated test still passing after changes: 🙂 

```
make testacc TESTARGS="-run=TestAccAWSSSMDocument_package"==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -count 1 -parallel 20 -run=TestAccAWSSSMDocument_package -timeout 120m
?       github.com/terraform-providers/terraform-provider-aws   [no test files]
=== RUN   TestAccAWSSSMDocument_package
=== PAUSE TestAccAWSSSMDocument_package
=== CONT  TestAccAWSSSMDocument_package
--- PASS: TestAccAWSSSMDocument_package (140.09s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       140.936s
```
